File: ui/packages/domain_models/lib/src/delivery.dart
import 'enums.dart';

class LocationCoord {
  final double latitude;
  final double longitude;

  const LocationCoord({
    required this.latitude,
    required this.longitude,
  });

  factory LocationCoord.fromJson(Map<String, dynamic> json) {
    return LocationCoord(
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {
    'latitude': latitude,
    'longitude': longitude,
  };
}

class RiderLocationPoint {
  final double latitude;
  final double longitude;
  final double? accuracy;
  final double? heading;
  final double? speed;
  final DateTime timestamp;

  const RiderLocationPoint({
    required this.latitude,
    required this.longitude,
    this.accuracy,
    this.heading,
    this.speed,
    required this.timestamp,
  });

  factory RiderLocationPoint.fromJson(Map<String, dynamic> json) {
    return RiderLocationPoint(
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      accuracy: json['accuracy'] != null ? (json['accuracy'] as num).toDouble() : null,
      heading: json['heading'] != null ? (json['heading'] as num).toDouble() : null,
      speed: json['speed'] != null ? (json['speed'] as num).toDouble() : null,
      timestamp: DateTime.parse(json['timestamp'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
    'latitude': latitude,
    'longitude': longitude,
    if (accuracy != null) 'accuracy': accuracy,
    if (heading != null) 'heading': heading,
    if (speed != null) 'speed': speed,
    'timestamp': timestamp.toIso8601String(),
  };
}

class DeliverySnapshot {
  final String deliveryId;
  final String orderId;
  final OrderStatus status;
  final String? riderId;
  final String? riderName;
  final String? riderPhone;
  final LocationCoord? storeLocation;
  final LocationCoord? destinationLocation;
  final RiderLocationPoint? lastRiderLocation;
  final int? etaSeconds;
  final double? remainingDistanceMeters;

  const DeliverySnapshot({
    required this.deliveryId,
    required this.orderId,
    required this.status,
    this.riderId,
    this.riderName,
    this.riderPhone,
    this.storeLocation,
    this.destinationLocation,
    this.lastRiderLocation,
    this.etaSeconds,
    this.remainingDistanceMeters,
  });

  factory DeliverySnapshot.fromJson(Map<String, dynamic> json) {
    return DeliverySnapshot(
      deliveryId: json['delivery_id'] as String,
      orderId: json['order_id'] as String,
      status: OrderStatus.fromString((json['status'] as String?) ?? 'pending'),
      riderId: json['rider_id'] as String?,
      riderName: json['rider_name'] as String?,
      riderPhone: json['rider_phone'] as String?,
      storeLocation: json['store_location'] != null
          ? LocationCoord.fromJson(json['store_location'] as Map<String, dynamic>)
          : null,
      destinationLocation: json['destination_location'] != null
          ? LocationCoord.fromJson(json['destination_location'] as Map<String, dynamic>)
          : null,
      lastRiderLocation: json['last_rider_location'] != null
          ? RiderLocationPoint.fromJson(json['last_rider_location'] as Map<String, dynamic>)
          : null,
      etaSeconds: json['eta_seconds'] as int?,
      remainingDistanceMeters: json['remaining_distance_meters'] != null
          ? (json['remaining_distance_meters'] as num).toDouble()
          : null,
    );
  }
}

File: backend/app/schemas/delivery.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict
from app.schemas.common import OrderStatus


class LocationCoord(BaseModel):
    latitude: float
    longitude: float


class RiderLocationPoint(BaseModel):
    latitude: float
    longitude: float
    accuracy: Optional[float] = None
    heading: Optional[float] = None
    speed: Optional[float] = None
    timestamp: datetime


class RiderLocationBatchRequest(BaseModel):
    locations: List[RiderLocationPoint]


class DeliveryStatusUpdateRequest(BaseModel):
    status: OrderStatus
    cod_collected: bool = False
    audit_reason: Optional[str] = None


class Delivery(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    order_id: str
    rider_id: Optional[str] = None
    rider_name: Optional[str] = None
    rider_phone: Optional[str] = None
    status: OrderStatus
    pickup_time: Optional[datetime] = None
    delivered_time: Optional[datetime] = None
    last_latitude: Optional[float] = None
    last_longitude: Optional[float] = None
    last_location_time: Optional[datetime] = None


class DeliverySnapshot(BaseModel):
    delivery_id: str
    order_id: str
    status: OrderStatus
    rider_id: Optional[str] = None
    rider_name: Optional[str] = None
    rider_phone: Optional[str] = None
    store_location: Optional[LocationCoord] = None
    destination_location: Optional[LocationCoord] = None
    last_rider_location: Optional[RiderLocationPoint] = None
    eta_seconds: Optional[int] = None
    remaining_distance_meters: Optional[float] = None
